### PR TITLE
JavaScript Connection implementation cleanup

### PR DESCRIPTION
**Pull request by pepone**
A small cleanup for JavaScript connection implementation

### CODE REVIEW

File: js/src/Ice/ConnectionI.js [hunk 1]
```diff
@@ -65,14 +65,13 @@ class MessageInfo {
...
+    constructor(communicator, instance, transceiver, endpoint, removeFromFactory, options) {
         this._communicator = communicator;
         this._instance = instance;
         this._desc = transceiver.toString();
         this._type = transceiver.type();
         this._endpoint = endpoint;
-        this._incoming = incoming;
-        this._adapter = adapter;
+        this._adapter = null;
         this._removeFromFactory = removeFromFactory;
```
Comment (pepone) [inline js/src/Ice/ConnectionI.js:75]: We only support outgoing connections in JavaScript. The adapter field is always null on construction for Outgoing connections.

File: js/src/Ice/ConnectionI.js [hunk 2]
```diff
@@ -737,7 +732,7 @@ export class ConnectionI {
             this._timer.cancel(this._closeTimeoutId);
         }
 
-        const traceLevels = this._instance.traceLevels();
+        const traceLevels = this._traceLevels;
```
Comment (pepone) [inline js/src/Ice/ConnectionI.js:735]: We cache _traceLevels and _logger for performance, but the cached fields were not used consistently.

File: js/src/Ice/ConnectionI.js [hunk 3]
```diff
@@ -1166,63 +1154,46 @@ export class ConnectionI {
     }
 
     validate() {
-        if (this._adapter !== null) {
```
Comment (pepone) [inline js/src/Ice/ConnectionI.js:1169]: This was checking if it was an incoming or outgoing connection. For JavaScript, we know it is always an outgoing connection.

File: js/src/Ice/OutgoingConnectionFactory.js [hunk 4]
```diff
@@ -254,8 +254,6 @@ export class OutgoingConnectionFactory {
                 this._instance,
                 transceiver,
                 endpoint.changeCompress(false).changeTimeout(-1),
-                false,
-                null,
                 (connection) => this.removeConnection(connection),
                 this._connectionOptions,
             );
```
Comment (pepone) [inline js/src/Ice/OutgoingConnectionFactory.js:259]: This is the only place were we create connections in JavaScript.

File: js/src/Ice/ConnectionI.js (hunk 1 above)
Comment (bernardnormier) [reply js/src/Ice/ConnectionI.js:75]: There is no null default in JS?

File: js/src/Ice/ConnectionI.js [hunk 5]
```diff
@@ -856,7 +851,7 @@ export class ConnectionI {
         const info = this._transceiver.getInfo();
         for (let p = info; p !== null; p = p.underlying) {
             p.adapterName = this._adapter !== null ? this._adapter.getName() : "";
-            p.incoming = this._incoming;
+            p.incoming = false;
```
Comment (bernardnormier) [inline js/src/Ice/ConnectionI.js:854]: Why do we need this incoming field, which I assume means whether the connection is incoming or outgoing?

File: js/src/Ice/ConnectionI.js (hunk 3 above)
Comment (bernardnormier) [reply js/src/Ice/ConnectionI.js:1169]: We did not check `this._incoming` but `this._adapter`. An outgoing connection can have an associated object adapter.

I assume the application can't get hold of the connection and call setAdapter before the call to `validate()`?

**Review by bernardnormier — APPROVED** [review]

File: js/src/Ice/ConnectionI.js (hunk 1 above)
Comment (pepone) [reply js/src/Ice/ConnectionI.js:75]: The default is **undefined**. The code uses [strict equality](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Strict_equality), which compares **null** and **undefined** as different values.

File: js/src/Ice/ConnectionI.js (hunk 5 above)
Comment (pepone) [reply js/src/Ice/ConnectionI.js:854]: This was part of the generated ConnectionInfo public API. I guess we can remove it for JavaScript in 3.8

**Pull request merged**